### ISSUE TITLE
Ensure all promises are runloop-aware to make it friendlier in acceptance tests

### ISSUE THREAD

**Issue by cibernox** [cibernox/ember-power-select#106]

**Comment by cibernox** [comment]
Apparently I fixed this without noticing it

**Comment by toranb** [comment]
I'm still getting a runloop error occasionally in acceptance tests using v0.5.2 and I'm curious if this fix was in the latest release or ?

**Comment by toranb** [comment]
Just a heads up about how/when this occurs - usually when we search (firing the performSearch func below) using something like debounce

```js
    //the find_all_stuff does a setter on search
    //and search is wrapped in a closure action like ...   search=(action "update_filter")
    update_filter(search) {
        Ember.run.debounce(this, this.get('find_all_stuff'), search, 300);
    }
```
here is the power-select code I see in the async stack

```js

    performSearch: function performSearch(term) {
      this.set('_searchText', term);
      if (this.get('search')) {
        this.performCustomSearch(term);
      } else {
        this.refreshResults();
      }
    }
```

**Comment by toranb** [comment]
Also - the top func in the stack is not the promise directly but instead "repositionDropdown" (performCustomSearch maybe the root cause)?

And this is the set that actually causes the auto-run to fail my acceptance tests

```js
this.set('_dropdownPositionClass', 'ember-basic-dropdown--above');
```

**Comment by toranb** [comment]
:) also it looks like the prop getting "set" that causes it ```_dropdownPositionClass```

**Comment by toranb** [comment]
It looks like even further ... basic dropdown might be the real issue. I see a dom listener for the "repositionDropdown" fired from ```insertBefore``` in ember v2.2 (stable)

Any chance the listener can be made run-loop aware or ?

**Comment by cibernox** [comment]
@toranb Is wrapping all the code in `repositionDropdown` in an `Ember.run(() => {  /* code here */ })` all that takes to make that event runloop-aware?

**Comment by cibernox** [comment]
Also, even if I don't think it's related, the way I was expecting the `search` action to work is by returning content from it, not by updating the `options`. 

It might work however, but I suspect that you're being exposed to race conditions by doing so.
Check this: http://www.ember-power-select.com/cookbook/debounce-searches

I think that you're approach is perfectly valid and I should check that if the return value from the `seach` action is undefined, don't wrap it into a promise: https://github.com/cibernox/ember-power-select/blob/master/addon/components/power-select/base.js#L184

**Comment by cibernox** [comment]
Since #137 the approach you where using for updating the list of options using the `search` action is supported. I don't think that it was related, but seems a valid approach to async searches too.

**Comment by toranb** [comment]
The approach we take is to let data get pushed down from the route when a query param changes. Updating the value in debounce simply updates the query param itself (allowing the refreshModel: true to do it's work).

Is your comment about line 184 mean a PR w/ this change would be good/accepted/etc ?

**Comment by cibernox** [comment]
@toranb I already made that change in PR #137. Your approach seems legit, but you need to make sure that `find_all_stuff` updates the options. And if you care about the `Loading ...` message, options must be a promise. 

I've just released `0.6.0-beta.1` with those changes. It contains a couple breaking changes if you're customizing the `selectedComponent` or `optionsComponent`, but the public API remains the same.

**Comment by toranb** [comment]
Excellent! thanks again for the quick release cycle(s)
